FocusOn/ViewModel/ProgressViewModel.swift

```diff
@@ -0,0 +1,125 @@
+    let bigProgress = Legend(color: .yellow, label: "Big Progress", order: 4)
+    let success = Legend(color: .green, label: "Success", order: 5)
+    
+    init( dataService: DataServiceProtocol = MockDataService()) {
+        self.dataService = dataService
+        self.allGoals = dataService.allGoals
+    }
+    
+    func fillChartData() -> [[DataPoint]]? {
+        guard let unwrappedGoals = allGoals else {
+            return nil
+        }
+        let weeklyChunksOfDataPoints = generateDataPointsForWeeklyChunksOf(goals: unwrappedGoals)
+        
+        return weeklyChunksOfDataPoints.isEmpty ? nil : weeklyChunksOfDataPoints
+    }
+    
+    private func calculateNumberOfCompletedTasks(goal: Goal) -> Int {
+        var numberOfCompletedTasks = 0
+        
+        for task in goal.tasks where task.isCompleted {
+            numberOfCompletedTasks += 1
+        }
+        return numberOfCompletedTasks
+    }
+    
+    private func calculateGoalProgress(goal: Goal) -> Legend {
+        let legendSignature: Legend
+        
+        switch calculateNumberOfCompletedTasks(goal: goal) {
+        case 0:
+            legendSignature = fail
+        case 1:
+            legendSignature = smallProgress
+        case 2:
+            legendSignature = bigProgress
+        case 3:
+            legendSignature = success
+        default:
+            legendSignature = noGoal
+        }
+        return legendSignature
+    }
+    
+    private func generateDataPointForGoal(goal: Goal) -> DataPoint {
+        let barHeight = Double(calculateNumberOfCompletedTasks(goal: goal) + 1)
+        let dayComponent = Calendar.current.component(.day, from: goal.createdAt)
+        let label = LocalizedStringKey(String(dayComponent))
+        let legend = calculateGoalProgress(goal: goal)
+        
+        return DataPoint(value: barHeight, label: label, legend: legend)
+    }
+    
+    private func splitGoalsIntoWeeklyChunks(goals: [Goal]) -> [[Goal]] {
+        let sortedGoals = goals.sorted { $0.createdAt < $1.createdAt }
+        var goalChunks: [[Goal]] = []
+        
+        for goal in sortedGoals {
+            let weekday = calendar.component(.weekday, from: goal.createdAt)
+            if weekday == firstWeekday || goalChunks.isEmpty {
+                goalChunks.append([goal])
+            } else {
+                goalChunks[goalChunks.count - 1].append(goal)
+            }
+        }
+        return goalChunks
+    }
+    
+    private func generateDataPointsForWeeklyChunksOf(goals: [Goal]) -> [[DataPoint]] {
+        let goalChunks = splitGoalsIntoWeeklyChunks(goals: goals)
+        var dataPointChunks: [[DataPoint]] = []
+        
+        for weeklyGoals in goalChunks {
+            var dataPointsForWeek: [DataPoint] = []
+            
+            guard let earliestDate = weeklyGoals.map({ $0.createdAt }).min() else {
+                continue
+            }
+            
+            // Find the first Monday in the week
```

Comment:
I think this part is not necessary, you just want to add missing data regardless of the position